### ISSUE TITLE
[STEP] step28 진행 - 함수 최적화

### ISSUE THREAD

**Issue by ghjang** [ghjang/deep-learning-from-scratch-3#35]
## 📌 step28 진행 추적 — 함수 최적화

**고지**: 제 3 고지 (고차 미분 계산) 4번째 step
**정확한 책 제목**: 함수 최적화 (브로 제공, 2026-08-19)
**제목 정정**: 기존 등록 "함수 최적화 (경사하강법)"에서 AI 부연설명 괄호 제거 → "함수 최적화"

### 🎯 이 step의 목표

1. **Rosenbrock 함수** 구현 — 최적화 벤치마크 클래식 ("바나나 함수", 전역 최소 (1,1)에서 y=0)
2. **경사하강법 (gradient descent) 갱신 루프** — 기울기 반대 방향으로 lr만큼 반복 이동
3. **반복 학습에서의 Define-by-Run** — 매 iteration 그래프가 새로 생성/폐기되는 구조 이해

### 📖 정답지 개요 (steps/step28.py — 인접 비교 완료)

```python
def rosenbrock(x0, x1):
    y = 100 * (x1 - x0 ** 2) ** 2 + (x0 - 1) ** 2

x0 = Variable(np.array(0.0)); x1 = Variable(np.array(2.0))
lr = 0.001; iters = 1000

for i in range(iters):
    print(x0, x1)
    y = rosenbrock(x0, x1)
    x0.cleargrad(); x1.cleargrad()   # grad 누적 방지 (step14)
    y.backward()
    x0.data -= lr * x0.grad          # 파라미터 갱신 (in-place)
    x1.data -= lr * x1.grad
```

### 🔑 핵심 개념 / 학습 포인트

1. **경사하강법** — loss가 가장 가파르게 감소하는 방향 (-gradient)으로 조금씩 (lr) 이동. 딥러닝 학습의 근간.
2. **학습률 (learning rate)** — 스텝 크기. 너무 크면 발산, 너무 작으면 수렴 느림. lr=0.001.
3. **`clear_grad` 매 iteration** — grad는 누적되도록 설계됨 (step14, 같은 Variable 여러 입력) → 갱신 루프에선 매번 초기화 필수.
4. **`x.data -= lr * x.grad`** — Variable.data (ndarray) 직접 in-place 갱신. 그래프에 연결 안 되는 "파라미터 업데이트" 경로.
5. **Define-by-Run의 반복 학습** — 매 iteration `y = rosenbrock(...)` 호출 시점에 그래프 새로 구축. 이전 그래프는 참조 끊기면 GC. "그래프가 영구적이지 않다"는 관점.
6. **Rosenbrock** — 비볼록 벤치마크. 골짜기가 굽어있어(바나나 모양) 경사하강법이 느리게 수렴하는 걸 보여주는 클래식.

### 📍 코드 위치 계획

- `rosenbrock` → step 한정 벤치마크 함수 → `rezero/steps/step28.py`에만 (sphere/goldstein과 동일 취급)
- 갱신 루프 데모 → `rezero/steps/step28.py`
- v1 패키지 변경 없음 (기존 기능으로 전부 커버 — clear_grad, fill_grad, 연산자)

### 🎨 rezero 변형 포인트 (예정)

- `cleargrad()` → `clear_grad()` (항목 021 — snake_case 네이밍 일관성)
- `y.backward()` → `fill_grad(y)` (rezero 정체성)
- `is_simple_core` 스위치 부분 — rezero는 무관 (core 하나)

### 🧪 예상 검증 포인트

- (0, 2) 시작 → (1, 1) 근처로 수렴하는지 (1000 iters, lr=0.001이면 부분 수렴 — rosenbrock은 느리게 수렴)
- lr 변화에 따른 발산/느린 수렴 관찰
- clear_grad 빼먹으면 grad 누적으로 이상 동작 확인 (실험 케이스)

### 📝 진행 로그

- 2026-08-19: 학습 준비. 제목 정정 (경사하강법 괄호 제거). 정답지 인접 비교(step27/28/29) — step28이 코드 주인 확인. step29는 뉴턴 방법 예고.

**Comment by ghjang** [comment]
## ✅ step28 완료 — 함수 최적화

### 구현 요약

**코드** (`rezero/steps/step28.py` — v1 반영 없음, 응용 step):
- `rosenbrock` — step 한정 벤치마크 (v1 승격 X)
- `gd_step` — 갱신 헬퍼 (복붙 버그 방지 + Optional 가드. 재실행 검증으로 [4]의 `x1.data -= lr * x0.grad` 복붙 버그를 포착한 뒤 도입)
- 데모 4케이스: 기본 수렴(낙하→크롤링 궤적 + 골짜기 바닥 편차 출력) / lr 발산(0.01부터) / clear_grad 누락(nan) / sphere 대비(거리 13% vs 44%)

### ★ 이 step의 심장 (사전 도출 — 제도화 첫 적용)

- "미분 기계가 처음으로 '일'을 한다" — 27 step 인프라로 실제 문제(최소화) 해결
- 4단계 루프 = 신경망 학습의 원형 (PyTorch zero_grad/backward/step의 조상)
- 적합성 메모 (제도화 첫 적용): 볼록+조건수 양호+gradient 저렴 → 강력 / ill-conditioned 골짜기 → 기어감

### 파생 산출물

- **탐구 노트 27** — Rosenbrock 해부(골짜기 바닥 = 포물선, A/B항) + 시험 함수 디자인 = 역설계(벤치마크 계보) + NFL 정리(만능 알고리즘 없음) + 뉴턴 vs SGD 선택 지도
- **탐구 노트 28** — 방향미분(→내적→코사인 3단 유도로 "gradient가 가장 가파른 방향" 회수) + gradient = 방향미분의 생성자 + 스칼라장/벡터장 + 갱신식 해부(greedy/자동 보폭/lr|∇f|² 감소 담보/관성 없음)
- **pdoc API 문서 파이프라인** — `PYTHONPATH=. uv run pdoc rezero.v1 -o output/api_doc` (dev 의존성 + AGENTS.md 메모)
- **AGENTS.md 제도화 2건** — 감동/학습 포인트 사전 도출 + 적합성 메모 (step27~28 패턴 제도화, step28이 첫 적용)
- v1 docstring 정비 — 전제 테이블 블록화(pdoc Markdown 대응) + 한자 정정(同名)

### 검증

- ✅ 테스트 105개 통과 / pyright 0 에러 / 데모 4케이스 재실행 정상 / v1 한자 잔존 0건

step29 (뉴턴 방법)로 — 선택 지도의 다음 좌표.
